### ISSUE TITLE
Async function handler in Restify causes assertion error

### ISSUE THREAD

**Issue by hajime-eb** [newrelic/node-newrelic#1824]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

Assertion error when New Relic is enabled and asynchronous handler is used in Restify. Setting `NEW_RELIC_LOG_LEVEL=trace` did not generate trace logs.

```
$ NEW_RELIC_APP_NAME=test node -r newrelic server.js
node:assert:125
  throw new AssertionError(obj);
  ^

AssertionError [ERR_ASSERTION]: Handler [testAsyncFunction] is missing a third argument (the "next" callback) but is not an async function. Middleware handlers can be either async/await or callback-based.Callback-based (non-async) handlers should accept three arguments: (req, res, next). Async handler functions should accept maximum of 2 arguments: (req, res).
    at Chain.add (/var/app/node_modules/restify/lib/chain.js:79:16)
    at forEach (/var/app/node_modules/restify/lib/server.js:678:23)
    at Array.forEach (<anonymous>)
    at Server.use (/var/app/node_modules/restify/lib/server.js:676:32)
    at Server.wrappedMounter [as use] (/var/app/node_modules/newrelic/lib/shim/webframework-shim.js:439:19)
    at Object.<anonymous> (/var/app/routes/createApiServer.js:9:24)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Module._load (node:internal/modules/cjs/loader:960:12) {
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: 'Function',
  expected: 'AsyncFunction',
  operator: '=='
}

Node.js v18.17.1
```

[NOTE]: # ( Describe the problem you're encountering. )
[TIP]:  # ( Do NOT share sensitive information, whether personal, proprietary, or otherwise! )

## Expected Behavior

Do not cause assertion error and can start listening.

[NOTE]: # ( Tell us what you expected to happen. )

## [Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) results

NA

[NOTE]: # ( Provide any other relevant log data. )
[TIP]:  # ( Scrub logs and diagnostic information for sensitive information )

## Steps to Reproduce

Please run the following command `NEW_RELIC_APP_NAME=test NEW_RELIC_LOG_LEVEL=trace node -r newrelic server.js`.

- `server.js` content
  ```js
  const restify = require("restify");
  
  async function testAsyncFunction(req, res) {
    console.log(req, res);
  }
  
  restify.createServer().use(testAsyncFunction).listen(12345, () => {});
  ```

When `-r newrelic` was deleted, no assertion error was displayed.

[NOTE]: # ( Please be as specific as possible. )
[TIP]:  # ( Link a sample application that demonstrates the issue. )

## Your Environment

I have confirmed this with the following versions.

- OS: macOS sonoma and debian bullseye on docker
- node: 18.17.1
- restify: 11.1.0
- newrelic: 11.2.1

[TIP]:  # ( Include as many relevant details about your environment as possible. )

## Additional context

[TIP]:  # ( Add any other context about the problem here. For example, relevant community posts or support tickets. )

NA

**Comment by mrickard** [comment]
@hajime-eb According to the Restify source, async handlers require an async function as the third parameter, and if that's not present (as in your repro), the function shouldn't be async. Supplying a function as a third parameter makes this work as expected. See https://github.com/restify/node-restify/blob/master/lib/chain.js#L78.

**Comment by bpyser** [comment]
@mrickard as the Assertion Error says, Restify now accepts middleware handlers with 2 different shapes.
>Middleware handlers can be either async/await or callback-based.Callback-based (non-async) handlers should accept three arguments: (req, res, next). Async handler functions should accept maximum of 2 arguments: (req, res).

The example that @hajime-eb has given is valid as of restify 9 and uses the: https://github.com/restify/node-restify/commit/12be9e243a407eaf7a30cbb16e399ee2a46dec93

Is there any plan for newrelic to support this updated route handler pattern?

**Comment by mrickard** [comment]
@bpyser @hajime-eb We can reopen this, though there are several priorities ahead of it in our roadmap. Of course, New Relic does welcome pull requests.

**Comment by mrickard** [comment]
Re-labeling as a feature request, as it's a request to support a recently-added mode of operation in restify.

**Comment by bpyser** [comment]
I'll try to put up a PR for this soon but in case this is helpful for anyone else, here is the patch-package we're currently testing:

**newrelic+10.6.2.patch**

```
diff --git a/node_modules/newrelic/lib/instrumentation/restify.js b/node_modules/newrelic/lib/instrumentation/restify.js
index 5cb8698..2096c49 100644
--- a/node_modules/newrelic/lib/instrumentation/restify.js
+++ b/node_modules/newrelic/lib/instrumentation/restify.js
@@ -57,6 +57,7 @@ module.exports = function initialize(agent, restify, moduleName, shim) {
         }
         return shim.recordMiddleware(middleware, {
           matchArity: true,
+          matchConstructorAsync: true,
           route,
           req: shim.FIRST,
           next: shim.LAST
@@ -72,6 +73,7 @@ module.exports = function initialize(agent, restify, moduleName, shim) {
       }
       return shim.recordMiddleware(middleware, {
         matchArity: true,
+        matchConstructorAsync: true,
         req: shim.FIRST,
         next: shim.LAST
       })
diff --git a/node_modules/newrelic/lib/shim/shim.js b/node_modules/newrelic/lib/shim/shim.js
index 2cb3707..454d0dc 100644
--- a/node_modules/newrelic/lib/shim/shim.js
+++ b/node_modules/newrelic/lib/shim/shim.js
@@ -1863,6 +1863,16 @@ function _wrap(shim, original, name, spec, args) {
   // Apply the spec and see if it returned a wrapped version of the property.
   let wrapped = spec.wrapper.apply(null, specArgs)
   if (wrapped && wrapped !== original) {
+    
+    if (spec.matchConstructorAsync && shim.isFunction(wrapped)) {
+      if (original.constructor.name === "AsyncFunction" && wrapped.constructor.name !== "AsyncFunction") {
+        const nonAsyncWrap = wrapped
+        wrapped = async function asyncWrapper() {
+          return nonAsyncWrap.apply(this, arguments)
+        }
+      }
+    }
+
     if (spec.matchArity && shim.isFunction(wrapped)) {
       wrapped = arity.fixArity(original, wrapped)
     }
diff --git a/node_modules/newrelic/lib/shim/specs/index.js b/node_modules/newrelic/lib/shim/specs/index.js
index cb06ef2..7aa7b8d 100644
--- a/node_modules/newrelic/lib/shim/specs/index.js
+++ b/node_modules/newrelic/lib/shim/specs/index.js
@@ -45,6 +45,7 @@ function cast(Class, spec) {
 function WrapSpec(spec) {
   this.wrapper = typeof spec === 'function' ? spec : spec.wrapper
   this.matchArity = hasOwnProperty(spec, 'matchArity') ? spec.matchArity : false
+  this.matchConstructorAsync = hasOwnProperty(spec, 'matchConstructorAsync') ? spec.matchConstructorAsync : false
 }
 
 function SegmentSpec(spec) {
diff --git a/node_modules/newrelic/lib/shim/webframework-shim.js b/node_modules/newrelic/lib/shim/webframework-shim.js
index 6a75f0f..2884d0f 100644
--- a/node_modules/newrelic/lib/shim/webframework-shim.js
+++ b/node_modules/newrelic/lib/shim/webframework-shim.js
@@ -1090,7 +1090,7 @@ function _isError(shim, err) {
  *   The spec object the values are coming from
  */
 function _copyExpectedSpecParameters(destination, source) {
-  const keys = ['matchArity']
+  const keys = ['matchArity', 'matchConstructorAsync']
 
   for (let i = 0; i < keys.length; ++i) {
     const key = keys[i]

```

**Comment by bizob2828** [comment]
@hajime-eb this was released in [v11.7.0](https://github.com/newrelic/node-newrelic/releases/tag/v11.7.0), please try again.
